### PR TITLE
Improve symiirorder2 performance

### PR DESCRIPTION
**Pull request by andfoy**
After benchmarking `symiirorder2` in its original version, several bottlenecks related to the starting conditions computation were found. This PR moves these computations into their own kernels and also tries (optimistically) to reduce the length of the linear search to be done when comparing against the `precision` argument, where it performs better when the cutoff is at the beginning of the array.

### CODE REVIEW

**Comment by andfoy** [comment]
The benchmark values before the proposed changes display a lower performance for all input sizes:


<details>
<summary>Benchmark script</summary>

```python
from itertools import product

import cupy as cp
import numpy as np
from cupy import testing
from cupyx.profiler import benchmark

from cupyx.scipy.signal import symiirorder2 as symiir2_gpu
from scipy.signal import symiirorder2 as symiir2_cpu

from tqdm import tqdm


input_sizes = [5, 10, 100, 1000, 5000, 10000, 50000, 100000, 200000, 500000,
               700000, 1_000_000, 2_000_000]
modules = [(cp, symiir2_gpu, 'CuPy'), (np, symiir2_cpu, 'SciPy')]
# kwargs = [('order', [1, 2, 4, 6, 10])]
kwargs = [('precision', [0.7049075329431513])]
# kwargs = [('extrapolate', [True])]
kwargs = list(product(*[list(product([x[0]], x[1])) for x in kwargs]))
measurement = {}


def gather_time(prof):
    cpu_time = prof.cpu_times.mean() * 1000
    gpu_time = prof.gpu_times.mean() * 1000
    return max(cpu_time, gpu_time)


def input_creation(xp, fn, size, precision=-1, **kwargs):
    size = size[0]
    x = testing.shaped_random((size,), xp, dtype=xp.float32, scale=0.1)
    omega = testing.shaped_random(
        (1,), xp, dtype=xp.float32, scale=2 * xp.pi)[0]
    r = testing.shaped_random((1,), xp, dtype=xp.float32, scale=0.3)[0]
    sig = (x, r, omega, precision)
    return fn, sig


def closure(fn, sig):
    # try:
    return fn(*sig)
    # except ValueError:
    #     pass


for kwarg_comb in kwargs:
    kwarg_measurement = measurement.get(kwarg_comb, {})
    measurement[kwarg_comb] = kwarg_measurement
    kwarg_comb = dict(kwarg_comb)
    for size in tqdm(input_sizes):
        size_measurement = kwarg_measurement.get(size, {})
        kwarg_measurement[size] = size_measurement
        for xp, cls, _id in modules:
            b, x = input_creation(xp, cls, (size,), **kwarg_comb)
            prof = benchmark(closure, (b, x), n_repeat=30, n_warmup=10)
            size_measurement[_id] = gather_time(prof)


lines = []
for kwarg_comb in kwargs:
    kwarg_line = ', '.join([f'{x}={y}' for x, y in kwarg_comb])
    lines.append(f'### `{kwarg_line}`\n')
    lines.append('| Size | CuPy (ms) | SciPy (ms) | Speedup |')
    lines.append('|:----:|:---------:|:----------:|:-------:|')
    kwarg_measurement = measurement[kwarg_comb]
    for size in input_sizes:
        comp = f'{size}'
        size_measurement = kwarg_measurement[size]
        times = []
        for _, _, _id in modules:
            time = size_measurement[_id]
            times.append(time)
            time_info = f'{time:3f}'
            comp = f'{comp} | {time_info}'
        speedup = times[1] / times[0]
        comp = f'{comp} | {speedup:3f}'
        lines.append(f'| {comp} |')
    lines.append('\n')

print('\n'.join(lines))
```

</details>

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 6.781028 | 0.002915 | 0.000430 |
| 10 | 6.618964 | 0.003675 | 0.000555 |
| 100 | 6.752166 | 0.004401 | 0.000652 |
| 1000 | 6.626852 | 0.014605 | 0.002204 |
| 5000 | 6.690649 | 0.053446 | 0.007988 |
| 10000 | 6.639221 | 0.100828 | 0.015187 |
| 50000 | 7.945073 | 0.546692 | 0.068809 |
| 100000 | 9.107076 | 0.955881 | 0.104960 |
| 200000 | 11.089818 | 1.899134 | 0.171250 |
| 500000 | 19.559299 | 4.801515 | 0.245485 |
| 700000 | 25.798957 | 6.664130 | 0.258310 |
| 1000000 | 35.142660 | 9.548301 | 0.271701 |
| 2000000 | 66.931391 | 19.130850 | 0.285828 |

**Comment by andfoy** [comment]
The benchmark after the proposed changes display an improvement when the input size increase, though they are comparable ith CPU speed:

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 5.847476 | 0.003812 | 0.000652 |
| 10 | 5.843630 | 0.003724 | 0.000637 |
| 100 | 5.804546 | 0.004210 | 0.000725 |
| 1000 | 5.915871 | 0.055692 | 0.009414 |
| 5000 | 5.784164 | 0.051283 | 0.008866 |
| 10000 | 5.657187 | 0.101061 | 0.017864 |
| 50000 | 5.746578 | 0.519686 | 0.090434 |
| 100000 | 5.655595 | 1.053095 | 0.186204 |
| 200000 | 5.961474 | 1.962874 | 0.329260 |
| 500000 | 6.084946 | 4.892054 | 0.803960 |
| 700000 | 6.571337 | 7.227331 | 1.099827 |
| 1000000 | 7.560214 | 9.616673 | 1.272011 |
| 2000000 | 11.722059 | 19.181492 | 1.636359 |

File: cupyx/scipy/signal/_splines.py [hunk 1]
```diff
@@ -149,13 +233,15 @@ def symiirorder2(input, r, omega, precision=-1.0):
...
     a3 = -rsq
     cs = cupy.atleast_1d(1 - 2 * r * cupy.cos(omega) + rsq)
+    omega = cupy.asarray(omega, cs.dtype)
+    precision *= precision
 
     # First compute the symmetric forward starting conditions
-    precision *= precision
     pos = cupy.arange(0, input.size + 2, dtype=input.dtype)
     diff = _compute_symiirorder2_fwd_hc(pos, cs, r, omega)
     err = diff * diff
```
Comment (andfoy) [inline cupyx/scipy/signal/_splines.py:247]: I tried to move the forward condition computation to its own kernel, but the performance dropped compared w.r.t current diff.

**Comment by andfoy** [comment]
After moving the forward starting condition to its own kernel ([c08901c](https://github.com/cupy/cupy/pull/7526/commits/c08901cb136ae23af9e9dc4f1cfb83b0110ff68c)):

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 5.506367 | 0.003520 | 0.000639 |
| 10 | 5.447737 | 0.003618 | 0.000664 |
| 100 | 5.397654 | 0.004325 | 0.000801 |
| 1000 | 5.442299 | 0.014540 | 0.002672 |
| 5000 | 5.370687 | 0.053203 | 0.009906 |
| 10000 | 5.536586 | 0.101152 | 0.018270 |
| 50000 | 5.374582 | 0.510151 | 0.094919 |
| 100000 | 5.519075 | 0.986211 | 0.178691 |
| 200000 | 5.713150 | 1.962043 | 0.343426 |
| 500000 | 6.297077 | 4.803445 | 0.762806 |
| 700000 | 6.650971 | 6.899020 | 1.037295 |
| 1000000 | 7.451553 | 9.655330 | 1.295747 |
| 2000000 | 7.905075 | 19.352731 | 2.448140 |

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by takagi** [comment]
Thanks, @andfoy!

**Pull request merged**